### PR TITLE
display whats new info in admin settings

### PR DESCRIPTION
**Pull request by blizzz**
contributes to #9163 by adding the what's new info to the admin notification

requires https://github.com/nextcloud/updater_server/pull/116

* [x] fix the CSS

@skjnldsv as you suggested I copied the corresponding popovermenu components from settings. However I needed to adjust the item a little bit for external links (target and rel). If it makes sense I can also port this over to the original (and don't hard code rel). Tell me.

Otherwise, the CSS styles do not match the linkless formats, and there's still a bit left to do to make it look beautiful. Some styles I applied. Currently:

screenshot_20180606_121204

So it definitely needs to get wider and perhaps some margins/paddings could get better. I cannot take over the styles completely from where it contains a link. Before i mess around more in the app, I guess it's better to get it back into apps.scss, too. 

Otherwise, first feedback @jancborchardt @skjnldsv ?

### CODE REVIEW

**Comment by blizzz** [comment]
Testing steps:

1. git clone the updater server https://github.com/nextcloud/updater_server and checkout `feature/115/whatsnew`
2. perhaps `composer install`, i don't recall…
3. patch it's config file to ship the info we want (see patch below)
4. start the updater server locally `php -S localhost:8888 index.php`
5. configure you dev instance to use the local updater server, per ` 'updater.server.url' => 'http://localhost:8888/',` in config.php
6. patch you dev instance to require these info (doesn't work with git channel), see path below
7. in apps/updatenotification run `make build-js`
8. go to /settings/admin/overview

patch for 3)
```diff
diff --git a/config/config.php b/config/config.php
index 594fa1b..74307cf 100755
--- a/config/config.php
+++ b/config/config.php
@@ -208,6 +208,22 @@ ZH1PxQilL7ihAvnOb2oU1Q==',
        ],
    ],
    'stable' => [
+       '14' => [
+           '100' => [
+               'latest' => '14.0.0',
+               'internalVersion' => '14.0.0.5',
+               'downloadUrl' => 'https://download.nextcloud.com/server/releases/nextcloud-14.0.0.zip',
+               'web' => 'https://docs.nextcloud.com/server/14/admin_manual/maintenance/upgrade.html',
+               'eol' => false,
+               'minPHPVersion' => '7.0',
+               'signature' => 'AG3egWHJkQaqCraVLiRq5Q7GvqArmBgVv8PNAfVYTOXPN6xWmw8cbUeBvBZ3OS1/
+bup1ktLDwTLDjWm1XMBUhhQCxVDLtEd3A6WfaVJz9BWoz1MyxUGlaIJSzTHLoEfz
+nFVUdoGH0mAdq8WtxRQSNwQWeKn+iF1jpbtIbDc29POtWvvzEgT/KW8MDzeidx6s
+W78tH4vldR0/aBn1D3hwnkQEQ8+Kz+Y4ItjHHi6XpJEfRQzYD0j9T+VTQ9IX4Xf/
+XqTrcaUCqwOlSC4pM7aUUzgaePPcYU2zrRDRaEgLma9eSkVMzkkc4kfM3izBG0Iv
+STb5hZFB2HMLyJxuj1l05w==',
+           ]
+       ],
        '13' => [
            '100' => [
                'latest' => '13.0.2',
```


patch for 6)
```diff
diff --git a/lib/private/Updater/VersionCheck.php b/lib/private/Updater/VersionCheck.php
index d34549f184..b4a28e8802 100644
--- a/lib/private/Updater/VersionCheck.php
+++ b/lib/private/Updater/VersionCheck.php
@@ -55,7 +55,7 @@ class VersionCheck {
     */
    public function check() {
        // Look up the cache - it is invalidated all 30 minutes
-       if (((int)$this->config->getAppValue('core', 'lastupdatedat') + 1800) > time()) {
+       if (false && ((int)$this->config->getAppValue('core', 'lastupdatedat') + 1800) > time()) {
            return json_decode($this->config->getAppValue('core', 'lastupdateResult'), true);
        }
 
@@ -70,7 +70,7 @@ class VersionCheck {
        $version = Util::getVersion();
        $version['installed'] = $this->config->getAppValue('core', 'installedat');
        $version['updated'] = $this->config->getAppValue('core', 'lastupdatedat');
-       $version['updatechannel'] = \OC_Util::getChannel();
+       $version['updatechannel'] = 'stable'; // \OC_Util::getChannel();
        $version['edition'] = '';
        $version['build'] = \OC_Util::getBuild();
        $version['php_major'] = PHP_MAJOR_VERSION;
```

**Comment by blizzz** [comment]
> P.S.: actually, you don't need to setup the updater_server. Instead patch VersionCheck::check() to deliver what we would have gotten. The updater_server make sense to test that one as well.

In that case, forget steps 1-5, start with 6 using this patch:

```diff
diff --git a/lib/private/Updater/VersionCheck.php b/lib/private/Updater/VersionCheck.php
index d34549f184..5b258ad055 100644
--- a/lib/private/Updater/VersionCheck.php
+++ b/lib/private/Updater/VersionCheck.php
@@ -54,6 +54,21 @@ class VersionCheck {
     * @return array|bool
     */
    public function check() {
+       return [
+           'version' => '14.0.0.5',
+           'versionstring' => 'Nextcloud 14.0.0',
+           'url' => 'https://download.nextcloud.com/server/releases/nextcloud-14.0.0.zip',
+           'web' => 'https://docs.nextcloud.com/server/14/admin_manual/maintenance/upgrade.html',
+           'changelog' => 'https://nextcloud.com/changelog/#14-0-0',
+           'whatsNew' =>
+               array (
+                   0 => 'Pollute the files failing to update l n files for pdf',
+                   1 => 'Emaillogin server show open graph preview in gallery removed old code',
+                   2 => 'Server sharee email search server log exceptions that s settings server',
+               ),
+           'autoupdater' => '1',
+           'eol' => '0',
+       ];
        // Look up the cache - it is invalidated all 30 minutes
        if (((int)$this->config->getAppValue('core', 'lastupdatedat') + 1800) > time()) {
            return json_decode($this->config->getAppValue('core', 'lastupdateResult'), true);

```

**Comment by blizzz** [comment]
I pushed some pixels and updated the screenshot in the description

File: apps/updatenotification/js-src/components/popoverMenu/popoverItem.vue [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<template>
+	<li>
+		<!-- If item.href is set, a link will be directly used -->
+		<a @click="item.action" v-if="item.href" :href="(item.href) ? item.href : '#' " :target="(item.target) ? item.target : '' " rel="noreferrer noopener">
```
Comment (blizzz) [inline apps/updatenotification/js-src/components/popoverMenu/popoverItem.vue]: @skjnldsv this is copied from `settings/src/components/popoverMenu/popovermenuItem.vue` as suggested. I added `:target` and `rel`. What are your current plans of consolidating it? Would be good to not diverge already. should i port it back?

File: apps/updatenotification/js-src/components/popoverMenu/popoverItem.vue [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+			<p v-else-if="item.longtext">{{item.longtext}}</p>
+		</a>
+		<!-- If item.action is set instead, a button will be used -->
+		<button @click="item.action" v-else-if="item.action">
+			<span :class="item.icon"></span>
+			<span v-if="item.text">{{item.text}}</span>
+			<p v-else-if="item.longtext">{{item.longtext}}</p>
+		</button>
+		<!-- If item.longtext is set AND the item does not have an action -->
+		<span class="item-wrapper" v-else>
```
Comment (blizzz) [inline apps/updatenotification/js-src/components/popoverMenu/popoverItem.vue]: @skjnldsv and here I added that class, but that's probably stupid… 🤔

Comment (skjnldsv) [reply apps/updatenotification/js-src/components/popoverMenu/popoverItem.vue]: @blizzz why do you need this class?
Indeed, this should be standardised, not customed to the needs of a specific app. :)

**Comment by skjnldsv** [comment]
I need to check again the css, I'm currently sick, so unfortunately not ready.
I'll take a look this afternoon, but like the other day, I couldn't get the notification to appear! :/

**Comment by blizzz** [comment]
> I need to check again the css, I'm currently sick, so unfortunately not ready.
I'll take a look this afternoon, but like the other day, I couldn't get the notification to appear! :/

Get well soon and get well first!

File: apps/updatenotification/js-src/components/popoverMenu/popoverItem.vue (hunk 2 above)
Comment (blizzz) [reply apps/updatenotification/js-src/components/popoverMenu/popoverItem.vue]: css style was not applied, adding this case to apps.scss fixed it, and i could get rid of this class and some rules: https://github.com/nextcloud/server/pull/9705/commits/67ec8443cefd09acb1280b703591150994a8ae83

**Review by skjnldsv — CHANGES_REQUESTED** [review]

**Review by skjnldsv — APPROVED** [review]

**Comment by skjnldsv** [comment]
CI is crying @blizz :)

**Comment by blizzz** [comment]
yeah, we also had some discussions today, cf. https://github.com/nextcloud/updater_server/pull/116#issuecomment-395434051 f, so some things gonna change that need to be reflected on client side.

File: apps/updatenotification/js-src/app.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+/**
```
[the remark this answers is not in the document]
Comment (blizzz) [reply apps/updatenotification/js-src/app.js]: stray from rebase i guess

**Comment by blizzz** [comment]
@skjnldsv i rebased, though all was good, psuhed, but now locally i see 

```
merged.js%20line%20774%20%3E%20eval:1:7

Error: Module build failed: Error: 

Vue packages version mismatch:

- vue@2.5.16
- vue-template-compiler@2.5.13

…

    at Object.<anonymous> (/srv/http/nextcloud/master/apps/notifications/node_modules/vue-template-compiler/index.js:8:9)
…


```

no matter how often i clean, npm install or update, removing the packages lock file, or make build-js things :( the merged.js is rebuilt an the package.json looks alright.

**Comment by skjnldsv** [comment]
@blizzz You need to clean your dev environment then :) 
Just do a make is enough and will install&compile from scratch

**Comment by blizzz** [comment]
> @blizzz You need to clean your dev environment then :)

@skjnldsv did it several times :(

**Comment by jancborchardt** [comment]
Looks very nice! :) But we should just use a simple ul (just round bullet points) instead of the stars. As it’s a bit confusing since normally we use the star for favoriting and it’s interactive.

(Also @skjnldsv here’s another case of why the default of the popover should be to have it centered ;)

**Comment by skjnldsv** [comment]
@jancborchardt 
doit

**Comment by rullzer** [comment]
I took the liberty to rebase.
If needed I still have the original branch backedup.

File: apps/updatenotification/lib/Settings/Admin.php [hunk 4]
```diff
@@ -107,6 +114,48 @@ public function getForm(): TemplateResponse {
...
+
+	protected function getLangTrunk(string $lang):string {
+		$pos = strpos($lang, '_');
+		if($pos !== false) {
+			$lang = substr($lang, 0, $pos);
+		}
+		return $lang;
+	}
+
+	protected function findWhatsNewTranslation(string $lang, array &$result, array $whatsNew): bool {
```
Comment (blizzz) [inline apps/updatenotification/lib/Settings/Admin.php]: for the user facing stuff i need to move it somewhere else. in the moment i am not sure where, so it is here.

File: lib/public/L10N/IFactory.php [hunk 5]
```diff
@@ -60,4 +60,14 @@ public function findAvailableLanguages($app = null);
...
+
+	/**
+	 * iterate through language settings (if provided) in this order:
+	 * 1. returns the forced language or:
+	 * 2. returns the user language or:
+	 * 3. returns the system default language or:
+	 * 4+∞. returns 'en'
+	 * @since 14.0.0
+	 */
+	public function iterateLanguage(bool $reset = false): string;
```
Comment (blizzz) [inline lib/public/L10N/IFactory.php]: that'll need to be mentioned in the 14 changes thing

File: core/Migrations/Version14000Date20180626223656.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+namespace OC\Core\Migrations;
+
+use OCP\DB\ISchemaWrapper;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version14000Date20180626223656 extends SimpleMigrationStep {
+	public function changeSchema(\OCP\Migration\IOutput $output, \Closure $schemaClosure, array $options) {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+		if(!$schema->hasTable('whats_new')) {
```
Comment (MorrisJobke) [inline core/Migrations/Version14000Date20180626223656.php]: The version.php needs to be raised to trigger this migration. And this should be added to the install routine as well. @rullzer @nickvergessen Or am I wrong?

Comment (MorrisJobke) [reply core/Migrations/Version14000Date20180626223656.php]: Version bump seems to be enough. Let me add it.

**Comment by MorrisJobke** [comment]
Let me rebase to fix the conflict.

**Review by MorrisJobke — APPROVED** [review]
Works 👍 

Let's fix the style in a followup PR

bildschirmfoto 2018-06-29 um 09 07 28

File: core/Migrations/Version14000Date20180626223656.php (hunk 6 above)
Comment (blizzz) [reply core/Migrations/Version14000Date20180626223656.php]: Right, thank you!

**Comment by MorrisJobke** [comment]
> Let's fix the style in a followup PR

#10047

**Pull request merged**